compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt

```diff
@@ -698,32 +689,75 @@ internal class DefaultFlingBehavior(
         lastAnimationCycleCount = 0
         // come up with the better threshold, but we need it since spline curve gives us NaNs
         return withContext(motionDurationScale) {
             if (abs(initialVelocity) > 1f) {
                 var velocityLeft = initialVelocity
                 var lastValue = 0f
-                AnimationState(
-                    initialValue = 0f,
-                    initialVelocity = initialVelocity,
-                ).animateDecay(flingDecay) {
-                    val delta = value - lastValue
-                    val consumed = scrollBy(delta)
-                    lastValue = value
-                    velocityLeft = this.velocity
-                    // avoid rounding errors and stop if anything is unconsumed
-                    if (abs(delta - consumed) > 0.5f) this.cancelAnimation()
-                    lastAnimationCycleCount++
+
+                // If this value is not null by the end of decayAnimation, it means that not entire provided
+                // delta was consumed during last animation frame, so the animation needs to be cancelled
+                // and [flingIntoOverscrollEffect] animation should be played
+                var unconsumedDeltaAfterDecay: Float? = null
+
+                // There is an edge case when a user overscrolls and slightly flings in direction of content
+                // but inertia is not enough to cover overscroll offset
+                // In that scenario, don't do decay animation and simply replace it with FlingIntoOverscrollEffect animation immediately
+                var needsDecayAnimation = true
+
+                if (flingIntoOverscrollEffect != null) {
+                    val targetValue = flingDecay.calculateTargetValue(0f, initialVelocity)
+
+                    if (flingIntoOverscrollEffect.isFlingInertiaWeak(targetValue)) {
+                        needsDecayAnimation = false
+                        unconsumedDeltaAfterDecay = 0f
+                    }
+                }
+
+                if (needsDecayAnimation) {
+                    AnimationState(
+                        initialValue = 0f,
+                        initialVelocity = initialVelocity,
+                    ).animateDecay(flingDecay) {
+                        val delta = value - lastValue
+                        val consumed = scrollBy(delta)
+                        lastValue = value
+                        velocityLeft = this.velocity
+
+                        val unconsumedDelta = delta - consumed
+
+                        // If some delta is not consumed, it means that fling hits into content bounds.
+                        // Unconsumed delta and current velocity will be initial values for
+                        // [flingIntoOverscrollEffect] animation to play if any, after we cancel decay animation
+                        if (abs(unconsumedDelta) > 0.5f) {
```

Comment:
Please reuse `Float.isAboutZero()`

Reply:
This particular line is reused from previous version of the code. I have a feeling that changing it could break things.

Reply:
It's simple one. It wasn't used because I added `isAboutZero` recently and didn't change unrelated code

Reply:
Oh, that's interesting. I thought isAboutZero to be about some really small epsilon. And was about to tell, that I didn't investigate accumulated error during delta consumption, so in case it doesn't fit that strict epsilon comparison, we could end up with having one extra frame of animation, where nothing visible happens (visual stutter from user experience side). But since it's exactly the same, I'd consider renaming `isAboutZero` to something like `isInsignificantForAnimation` and dual `isSignificantForAnimation` or at at least adding some comment to it. Which one you think is preferable?

Reply:
Let's avoid negation ("in") in names if possible